x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java

```diff
@@ -354,32 +355,30 @@ protected void onListShards(List<SearchShard> shards, List<SearchShard> skipped,
             // best effort to cancel expired tasks
             checkCancellation();
             searchResponse.compareAndSet(null,
-                new MutableSearchResponse(shards.size() + skipped.size(), skipped.size(), clusters,
-                    aggReduceContextSupplier, threadPool.getThreadContext()));
+                new MutableSearchResponse(shards.size() + skipped.size(), skipped.size(), clusters, threadPool.getThreadContext()));
             executeInitListeners();
         }
 
         @Override
-        public void onPartialReduce(List<SearchShard> shards, TotalHits totalHits, InternalAggregations aggs, int reducePhase) {
+        public void onPartialReduce(List<SearchShard> shards, TotalHits totalHits,
+                DelayableWriteable.Serialized<InternalAggregations> aggregations, int reducePhase) {
             // best effort to cancel expired tasks
             checkCancellation();
-            searchResponse.get().updatePartialResponse(shards.size(),
-                new InternalSearchResponse(new SearchHits(SearchHits.EMPTY, totalHits, Float.NaN), aggs,
-                    null, null, false, null, reducePhase), aggs == null);
+            Supplier<InternalAggregations> reducedAggs = () -> aggregations == null ?
+                    null : InternalAggregations.topLevelReduce(singletonList(aggregations.get()), aggReduceContextSupplier.get());
```

Comment:
Why do we apply a top level reduce here ? We want it to be applied on demand (when the user gets a partial response) in order to avoid doing work when it's not needed. We should keep the pointer to the serialized aggs so that we don't double the memory in async search ?

Reply:
It *is* on demand via the `Supplier<> foo = () ->` part. At least, I thought it was.

Reply:
I understand now, the final reduce is a supplier. Although the `() -> aggregations == null` is hard to read, maybe rewrite it into `() -> (aggregations == null) ?`

Reply:
I'll rewrite it with good stuff like comments!